### PR TITLE
Add vCard plugin

### PR DESCRIPTION
**Pull request by ggozad**
This is intended to figure what needs to be done to help plugin authors.
Here's a simple vCard plugin. It can set a vCard from an object and get a vCard retuning an object.
Things I would like to figure out:

Given say a mock client such as:
```
import {Client} from '@xmpp/client-core'
import { vCard } from '@xmpp/protocols'

function mockClient() {
  const c = new Client()
  c._domain = 'bar'
  c._jid('foo@bar')
  return c
}
```

how would one fake the server? i.e. how do you inspect that the client sent a stanza to write expectations and how do you mock the client receiving a stanza from the server?

### CODE REVIEW

File: packages/plugins/vcard/vcard.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      return caller.set(null, _buildvCard(vcard))
+    },
+  }
+}
+
+export default {
+  name: 'vCard',
+  NS_VCARD,
+  plugin,
+}
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: We use `module.exports` for now but I'd like to switch to `.mjs` as soon as Node.js 8 gets released.

File: packages/plugins/vcard/vcard.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      })
+    },
+    set(vcard) {
+      return caller.set(null, _buildvCard(vcard))
+    },
+  }
+}
+
+export default {
+  name: 'vCard',
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: I find it easier if we always use lower case here

File: packages/plugins/vcard/vcard.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: `use strict`

File: packages/plugins/vcard/vcard.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    for (let [key, val] of Object.entries(dict)) {
+      if (typeof val === 'object')
+        builder.cnode(_buildvCard(val, xml`<${key}/>`)).up()
+      else if (val)
+        builder.c(key).t(val)
+      else
+        builder.c(key).up()
+    }
+    return builder
+  }
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: please move those functions outside of the `plugin` function and maybe export them so the user can use them and we can test them

**Comment by sonnyp** [comment]
I agree we'd need that but for this plugin and because it uses `iq-caller` underneath I feel that unit testing the build and parse functions only is best. What do you think?

File: packages/plugins/vcard/vcard.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  const _buildvCard = (dict, parent) => {
+    let builder = parent ? parent : xml`<vCard xmlns="${NS_VCARD}" version="2.0"/>`
+    for (let [key, val] of Object.entries(dict)) {
+      if (typeof val === 'object')
+        builder.cnode(_buildvCard(val, xml`<${key}/>`)).up()
+      else if (val)
+        builder.c(key).t(val)
+      else
+        builder.c(key).up()
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: I think the linter is gonna fail here, make sure that `make test` passes. You'll need prosody 0.10 installed (but no need to configure it) and stopped.

**Comment by ggozad** [comment]
Thank @sonnyp :) I am gonna be away for a couple of days so next week I will adapt to your remarks. With regards to the tests, I think it would be a good idea to at least have a documented way of testing plugins that works as a recipe even if these are not part of the package.

**Comment by sonnyp** [comment]
@ggozad just to let you know I'm working on the plugin architecture, I'm gonna make some change as the current 'API' was kinda temporary. I'll cc you on the PR to get your feedback.

**Comment by ggozad** [comment]
@sonnyp does that mean I should wait with this until you implement the changes you want?

**Comment by sonnyp** [comment]
@ggozad no.

I recommend we land this PR before the new plugin interface.

For testing, here is what I made so far based on your recommendation 

https://github.com/node-xmpp/node-xmpp/blob/master/packages/plugins/iq-callee/test.js#L16

It allows to mock the reception of stanza and assert the client's response.

File: packages/plugins/index.js [hunk 6]
```diff
@@ -8,7 +8,8 @@
   'ping',
   'stanza-router',
   'version',
-  'time'
+  'time',
+  'vcard',
 ].map(plugin => require(`./${plugin}`)).forEach((plugin) => {
   module.exports[plugin.name] = plugin
 })
```
Comment (sonnyp) [inline packages/plugins/index.js]: I removed this file, it was unnecessary.

File: packages/plugins/vcard/vcard.js (hunk 3 above)
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: I added the eslint rule

File: packages/plugins/vcard/vcard.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  return {
+    entity,
+    get(jid) {
+      return caller.get(jid,
+        xml`<vCard xmlns='${NS_VCARD}'/>`
+      )
+      .then((res) => {
+        return Promise.resolve(_parsevCard(res))
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: no need for Promise.resolve

`.then(res => _parsevCard(res))`

**Comment by ggozad** [comment]
For some reason lib/Plugin is missing after the rebase, probably related to the filename changes.

**Comment by sonnyp** [comment]
@ggozad I removed `lib/plugin.js` and moved the code to `index.js`

File: packages/plugins/index.js [hunk 8]
```diff
@@ -9,6 +9,7 @@
   'stanza-router',
   'version',
   'time',
+  'vcard',
```
Comment (sonnyp) [inline packages/plugins/index.js]: please remove this, I've removed the `packages/plugins/index.js` file

File: packages/plugins/vcard/vcard.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  return builder
+}
+
+function plugin (entity) {
+  const caller = entity.plugin(iqCaller)
+
+  return {
+    entity,
+    get (jid) {
+      return caller.get(jid, xml`<vCard xmlns='${NS_VCARD}'/>`)
```
Comment (sonnyp) [inline packages/plugins/vcard/vcard.js]: the API changed, `jid` is not an optional argument and moved to the second position which allow to use spread arguments

use

```javascript
get (...args) {
  return caller.get(xml`<vCard xmlns='${NS_VCARD}'/>`, ...args)
}
```
same for `set`

Comment (ggozad) [inline packages/plugins/vcard/vcard.js]: Hah no worries, will get to it :)

**Comment by ggozad** [comment]
I can't say I understand how to test this. Presumably I would want:
```
test('get', t => {
  t.context.plugin.get('foo@bar')
  .then((vcard) => {
  .... SECTION A....
   })  

}
```
In section A  I would want to inspect the vcard.
I am missing where I would inspect the stanza that the client sent and reply with a vcard. Could you provide more info?

**Comment by ggozad** [comment]
Added tests. I explicitly added an assertion for the jid repositioning which fails for me, know why?

**Comment by sonnyp** [comment]
> Added tests. I explicitly added an assertion for the jid repositioning which fails for me, know why?

My bad. https://github.com/node-xmpp/node-xmpp/blob/master/packages/plugins/iq-caller/index.js#L39

I forgot to pass `...args` down to `request`

Could you replace with `return this.request(iq, ...args)` in both `get` and `set` methods of the `iq-caller/index.js` file ?

**Comment by sonnyp** [comment]
@ggozad thanks, will add tests for iq-calleer.

File: packages/plugins/vcard/index.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    } else {
+      dict[c.name] = parsevCard(c)
+    }
+  })
+  return dict
+}
+
+const buildvCard = (dict, parent) => {
+  const builder = parent || xml`<vCard xmlns="${NS_VCARD}" version="2.0"/>`
+  for (const [key, val] of Object.entries(dict)) {
```
Comment (sonnyp) [inline packages/plugins/vcard/index.js]: travis is failing, :( `Object.entries` is Node.js >= 7 only 

We support Node.js >= 4

I'll configure eslint accordingly.

Comment (sonnyp) [inline packages/plugins/vcard/index.js]: hmm we use other features that are unsupported by Node.js 4, I'll fix this myself

**Comment by sonnyp** [comment]
@ggozad merged, thanks!

**Comment by ggozad** [comment]
pleasure, gonna move on with pubsub which is a lot more complex.

**Comment by ggozad** [comment]
Would it be hard to release the new API? I tried to install from local but failed on several packages.

**Comment by sonnyp** [comment]
@ggozad actually it's already in master, feel free to give it a try and let me know your thoughts about it if you do so.

See https://github.com/node-xmpp/node-xmpp/blob/master/packages/plugins/iq-caller/index.js#L1 for an example.

You simply call the plugin function with the plugin name as first argument, an object as second argument and an array of plugin dependencies as third argument. 

The `start` method is called when the plugin is attached to the client.
The `stop` method is called when the plugin is detached from the client.

Checkout https://github.com/node-xmpp/node-xmpp/tree/master/packages/plugin for the underlying implementation.

You could try to port the vcard plugin to it as a first step.

**Comment by ggozad** [comment]
Gonna have a look tomorrow and port vcard. I meant actually releasing on npm if you feel you have come close enough to stability.

**Comment by sonnyp** [comment]
@ggozad we could publish a non stable version but unless necessary I don't feel like it's worth the trouble before having more tests and documentation, I find working with the repository much easier.

**Pull request merged**